File: quarb-code/tests/fixtures/vocab.c
#include <stdio.h>

#define LIMIT 10
#define STEP(x) ((x) + 1)

struct lexer {
    int pos;
};

enum token {
    WORD,
    SPACE,
};

typedef int (*step_fn)(int);

static int *helper(int a, int b);

/* Advance to the limit. */
static int *helper(int a, int b)
{
    static int r;
    r = a;
    for (int i = 0; i < a; i++) {
        if (i > b) {
            r += STEP(i);
        } else if (i == b) {
            r -= i;
        }
    }
    while (r > LIMIT) {
        r--;
    }
    switch (r) {
    case 0:
        r = 1;
        break;
    default:
        r = 0;
        break;
    }
    return &r;
}

int main(void)
{
    int *p = helper(1, 2);
    printf("%d\n", *p);
    return p == 0 ? 1 : 0;
}

File: quarb-code/tests/fixtures/vocab.rs
mod lexer {
    /// Scan the input.
    pub fn lex(input: &str) -> Vec<char> {
        fn is_name_char(c: char) -> bool {
            c.is_alphanumeric() || c == '_'
        }
        let mut out = Vec::new();
        for c in input.chars() {
            if is_name_char(c) {
                out.push(c);
            } else if c == ' ' {
                out.push('_');
            } else {
                match c {
                    '\n' => out.push('.'),
                    _ => {}
                }
            }
        }
        out
    }
}

/// A cursor over the input.
struct Lexer {
    pos: usize,
}

enum Token {
    Word,
    Space,
}

impl Lexer {
    fn helper(&mut self, n: usize) -> usize {
        let f = |x: usize| x + 1;
        while self.pos < n {
            self.pos = f(self.pos);
        }
        self.pos
    }
}

fn run(n: usize) -> usize {
    let mut lx = Lexer { pos: 0 };
    lx.helper(n)
}

const LIMIT: usize = 10;

use std::fmt;
